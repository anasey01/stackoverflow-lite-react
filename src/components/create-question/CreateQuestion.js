import React, { Component } from 'react';
import { connect } from 'react-redux';
import propTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { notify, ToastContainer } from '../../utilities/toast/notify';
import { createQuestion } from '../../redux/actions/create-question/createQuestionActions';
import '../styles/createQuestion.scss';

const mapDispatchToProps = dispatch => ({
  createUserQuestion: question => dispatch(createQuestion(question))
});

export class ConnectedCreateQuestion extends Component {
  constructor(props) {
    super(props);
    this.state = {
      questionTitle: '',
      questionContent: '',
      questionId: null,
    };
  }

  handleChange = (event) => {
    this.setState({
      [event.target.name]: event.target.value,
    });
  };

  handleSubmit = async (event) => {
    event.preventDefault();
    const {
      createUserQuestion,
      history,
    } = this.props;
    const questionInfo = await createUserQuestion(this.state);
    const {
      success,
      message,
    } = questionInfo;
    switch (success) {
    case true:
      notify(message, 'success');
      // eslint-disable-next-line no-case-declarations
      const questionId = questionInfo.question.question_id;
      this.setState({
        questionId,
      });
      break;
    default:
      return notify(message, 'failure');
    }
    return setTimeout(() => history.push(`/questions/${this.state.questionId}`), 1000);
  };

  render() {
    return (
      <div className="main">
        <aside>
          <div id="left-sidebar">
            <nav>
              <ul>
                <li><Link to="/">Home</Link></li>
                <li><Link to="/recentquestion">Recent Questions</Link></li>
                <li id="no-border" className="highlight-current"><Link to="/createquestion">Ask a Question</Link></li>
              </ul>
            </nav>
          </div>
        </aside>
        <div className="main-content-title">
          <form id="create-question-form" onSubmit={this.handleSubmit}>
            <ToastContainer />
            <label htmlFor="title">Title:</label>
            <input type="text"
              onChange={this.handleChange}
              id="title" name="questionTitle"
              placeholder="Title" />
            <label htmlFor="content">Your Question:</label>
            <textarea name="questionContent"
              onChange={this.handleChange}
              id="content"
              placeholder="Your Question Here..." />
            <button type="submit"
              className="btn question-answer-btn"
              id="btn"> Post Your Question </button>
          </form>
        </div>
      </div>
    );
  }
}

ConnectedCreateQuestion.propTypes = {
  createUserQuestion: propTypes.func.isRequired,
  history: propTypes.object,
};

const CreateQuestion = connect(null, mapDispatchToProps)(ConnectedCreateQuestion);
export default CreateQuestion;
