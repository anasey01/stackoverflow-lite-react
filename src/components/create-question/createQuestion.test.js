import React from 'react';
import { mount, shallow } from 'enzyme';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import CreateQuestion, { ConnectedCreateQuestion } from './CreateQuestion';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

describe('<CreateQuestion />', () => {
  it('matches the snapshot', () => {
    const initialState = {};
    const store = mockStore(initialState);

    const wrapper = shallow(<Provider store={store}>
      <CreateQuestion />
    </Provider>);
    expect(wrapper).toMatchSnapshot();
  });

  it('should simulate state change', () => {
    const mockCreateQuestionFn = jest.fn();
    const wrapper = mount(<ConnectedCreateQuestion
      createUserQuestion={mockCreateQuestionFn} />);

    const title = wrapper.find('#title');
    title.instance().value = 'Some title';
    title.simulate('change');

    const content = wrapper.find('#content');
    content.instance().value = 'Some content';
    content.simulate('change');

    expect(wrapper.state()).toMatchObject({
      questionTitle: 'Some title',
      questionContent: 'Some content',
      questionId: null,
    });
    wrapper.unmount();
  });

  it('should simulate submiting form successfully', () => {
    const mockCreateQuestionFn = jest.fn();
    const wrapper = mount(<ConnectedCreateQuestion
      createUserQuestion={mockCreateQuestionFn} />);

    const title = wrapper.find('#title');
    title.instance().value = 'Some title';
    title.simulate('change');

    const content = wrapper.find('#content');
    content.instance().value = 'Some content';
    content.simulate('change');

    const form = wrapper.find('form');
    form.simulate('submit', { preventDefault() {} });
    expect(mockCreateQuestionFn.mock.calls[0][0]).toEqual({
      questionTitle: 'Some title',
      questionContent: 'Some content',
      questionId: null,
    });
    wrapper.unmount();
  });
});
