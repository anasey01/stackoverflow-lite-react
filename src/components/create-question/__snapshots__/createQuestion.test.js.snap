// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateQuestion /> matches the snapshot 1`] = `<Connect(ConnectedCreateQuestion) />`;
